test/integration/scheduler_perf/util.go

```diff
@@ -306,28 +308,74 @@ func (tc *throughputCollector) run(ctx context.Context) {
 		klog.Fatalf("%v", err)
 	}
 	lastScheduledCount := len(podsScheduled)
-	ticker := time.NewTicker(throughputSampleFrequency)
+	ticker := time.NewTicker(throughputSampleInterval)
 	defer ticker.Stop()
+	lastSampleTime := time.Now()
+	started := false
+	skipped := 0
+
 	for {
 		select {
 		case <-ctx.Done():
 			return
 		case <-ticker.C:
+			now := time.Now()
 			podsScheduled, err := getScheduledPods(tc.podInformer, tc.namespaces...)
 			if err != nil {
 				klog.Fatalf("%v", err)
 			}
 
 			scheduled := len(podsScheduled)
-			// Only do sampling if number of scheduled pods is greater than zero
-			if scheduled > 0 {
-				samplingRatioSeconds := float64(throughputSampleFrequency) / float64(time.Second)
-				throughput := float64(scheduled-lastScheduledCount) / samplingRatioSeconds
-				tc.schedulingThroughputs = append(tc.schedulingThroughputs, throughput)
-				lastScheduledCount = scheduled
-				klog.Infof("%d pods scheduled", lastScheduledCount)
+			// Only do sampling if number of scheduled pods is greater than zero.
+			if scheduled == 0 {
+				continue
+			}
+			if !started {
+				started = true
+				// Skip the initial sample. It's likely to be an outlier because
+				// sampling and creating pods get started independently.
+				lastSampleTime = now
+				continue
+			}
+
+			newScheduled := scheduled - lastScheduledCount
+			if newScheduled == 0 {
+				// Throughput would be zero for the interval.
+				// Instead of recording 0 pods/s, keep waiting
+				// until we see at least one additional pod
+				// being scheduled.
```

Comment:
I think the algorithm here is different from previous one, let's say we get a scheduling results as [10,0,0,0,20], then previous one will be [10, 20], but now, it will be [10,5,5,5,5], this seems not right.

Reply:
With `[10,0,0,0,20]`, you mean 10 pods in the first interval, 0 in the next three, then 20? You are right, the existing code [skips intervals with zero pods](https://github.com/kubernetes/kubernetes/blob/c0804231b97e418429f5f353c94e2c59db0d4c7a/test/integration/scheduler_perf/util.go#L323).

That's actually broken!

Let's assume each interval is one second long, so the scheduling took 5 seconds. The overall average is therefore 30 pods/5 seconds = 6 pods/s. When recording the average based on the sampled intervals, the code [sums up the value and divides by the number of samples](https://github.com/kubernetes/kubernetes/blob/c0804231b97e418429f5f353c94e2c59db0d4c7a/test/integration/scheduler_perf/util.go#L346), which leads to 15 pods/s - much too high.

I misremembered the existing code: I thought it started sampling once it gets the first interval with non-zero results (that's what I do now), but it didn't do that.

For this example, the new code should record `[10, 5, 5, 5, 5]`, as you said, which yields 30/5 = 6 pods/s when averaging the samples, which is the correct result.

Reply:
> [10,0,0,0,20] > [10, 20]

Yes, the average throughput is not right.

> [10,0,0,0,20] > [10, 5, 5, 5, 5]

This buries the truth that we do have low throughput and high throughput when calculating quantiles, like P50, P99.

So, I think both approaches have their problems, we used to ignore the 0 throughput because we start the calculation from creating the pods, then we have lots of slots with 0 throughput, as a result, we may get a P50 as 0, but now we will start our calculation when we get pods scheduled(am I right?), so I think it's ok to include the 0 throughput which might be a real problem. Any ideas? 
Also cc @alculquicondor

Reply:
The problem of "zero pods per interval" can also be described as "the intervals are too short for the current workload". Averaging over multiple intervals mitigates that. It's just a mitigation. The right solution would be to run each workload with an appropriate interval size - the problem just is that we don't know what's appropriate until after the test run.

Let's look at one example with interval 1 second and an average scheduling rate of 0.5 pods/s. Without averaging, we record [1, 0, 1, 0, 1, 0, 1]. I don't remember whether P50 rounds the index up or down, but it doesn't matter, because neither P50=0 nor P50=1 are useful.

With averaging, we record [1, 0.5, 0.5, 0.5, 0.5, 0.5, 0.5]. P50 becomes 0.5. This seems more useful to me.

Reply:
> the intervals are too short for the current workload

1 second I think is long enough for pod scheduling right now, and we're talking about something may rarely happen, so I'm ok with this change, can you help to fix the tests, I'm LGTM and leave this for other approvers for another around of review.

Reply:
Reporting P50=0 for a perfectly linear average scheduling rate of 0,5 pods/s doesn't seem useful to me. It's just an artifact of a sampling interval that is too small.